extensions/amp-mraid/0.1/amp-mraid.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Connects AMP Host Services to MRAID to allow invoking native
+ * APIs from mobile app ad webviews.
+ *
+ * Example:
+ * <code>
+ * <script async host-api="amp-mraid"
+ *  fallback-on="load incomplete"></script>
+ * </code>
+ *
+ */
+
+import {HostServices} from '../../../src/inabox/host-services';
+import {dev} from '../../../src/log';
+import {getMode} from '../../../src/mode';
+import {MraidService} from './mraid-service';
+
+const TAG = 'amp-mraid';
+const FALLBACK_ON = 'fallback-on';
+
+/**
+ * Error codes that can be macroed into the error ping url.
+ * @const @enum {string}
+ */
+const ErrorCodes = {
+  load: 'load',
+  incomplete: 'incomplete',
+}
+
+/**
+ * Loads mraid.js if available, and once it's loaded looks good, configures an
+ * MraidService to handle visibility, fullscreen, and exit.
+ */
+export class MraidInitializer {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {!Array<string>} */
+    this.fallbackOn_ = [];
```

Comment:
can you explain this fallback attribute a bit? I thought we're going to have inabox runtime decide what to fallback based on the error code.

Reply:
Sure!  The reason for making fallback-on something configurable is that at least in the Google case we want to have different fallback behavior for visibility in different circumstances.  Specifically:

* If we know we're serving to a mobile app, then we *never* want to fall back to `IntersectionObserver`.  Falling back in this case could result in categorizing creatives as visible the containing webview isn't even on screen.

* If we can't tell whether we're serving to a mobile app, then we want to use the existence of `mraid.js` to tell us which situation we're in.  If `mraid.js` is present (regardless of whether it's functional) then we don't want to fall back to `IntersectionObserver` for the reason above, but if it's absent then we're probably not in a mobile app and it's ok to fall back.

Reply:
Yes, sorry, "we" is the ad server, and the ad server needs to be able to tell the client how to handle failure.

As implemented in this PR, usage looks like any of:

    fallback-on=""
    fallback-on="load"
    fallback-on="incomplete"
    fallback-on="load incomplete"

That is, you can say what sort of errors you would like to fall back on.  There's:

* `load`:  `mraid.js` doesn't exist
* `incomplete`: `mraid.js` exists, but doesn't work

Google's ad server would pass `fallback-on=""` when serving to mobile apps, and `fallback-on="load"` when serving to something that might be a mobile app or might be a browser.

Reply:
I think it's better for runtime not to know about those fallback options, and let the extension communicating to runtime using standard error codes. Meaning those attributes are intended to be used by extension only. Would that work in your case?

Also, any other naming options you have in mind?

Reply:
this PR might give a better picture what I meant about using error code: #20357

Reply:
Ok, reworked in 5af581b7b